diff --git a/meson.build b/meson.build
index 997218d..6839409 100644
--- a/meson.build
+++ b/meson.build
@@ -319,6 +319,13 @@ elif host_machine.system() == 'ethereal'
 	internal_conf.set10('MLIBC_MMAP_ALLOCATE_DSO', false)
 	internal_conf.set10('MLIBC_MAP_FILE_WINDOWS', true)
 	subdir('sysdeps/ethereal')
+elif host_machine.system() == 'nyaux'
+	rtld_include_dirs += include_directories('sysdeps/nyaux/include')
+	libc_include_dirs += include_directories('sysdeps/nyaux/include')
+	internal_conf.set10('MLIBC_MAP_DSO_SEGMENTS', false)
+	internal_conf.set10('MLIBC_MMAP_ALLOCATE_DSO', false)
+	internal_conf.set10('MLIBC_MAP_FILE_WINDOWS', false)
+	subdir('sysdeps/nyaux')
 # ANCHOR: demo-sysdeps
 elif host_machine.system() == 'demo'
 	subdir('sysdeps/demo')

diff --git a/sysdeps/nyaux/crt1.S b/sysdeps/nyaux/crt1.S
new file mode 100644
index 0000000..b6964d9
--- /dev/null
+++ b/sysdeps/nyaux/crt1.S
@@ -0,0 +1,7 @@
+.section .text
+.global _start
+_start:
+	lea main(%rip), %rdi
+	mov %rsp, %rsi
+	call __mlibc_entry
+.section .note.GNU-stack,"",%progbits
diff --git a/sysdeps/nyaux/crti.S b/sysdeps/nyaux/crti.S
new file mode 100644
index 0000000..911b078
--- /dev/null
+++ b/sysdeps/nyaux/crti.S
@@ -0,0 +1,11 @@
+.section .init
+.global _init
+_init:
+	push %rax
+
+.section .fini
+.global _fini
+_fini:
+	push %rax
+.section .note.GNU-stack,"",%progbits
+
diff --git a/sysdeps/nyaux/crtn.S b/sysdeps/nyaux/crtn.S
new file mode 100644
index 0000000..0187e50
--- /dev/null
+++ b/sysdeps/nyaux/crtn.S
@@ -0,0 +1,9 @@
+.section .init
+	pop %rax
+	ret
+
+.section .fini
+	pop %rax
+	ret
+.section .note.GNU-stack,"",%progbits
+
diff --git a/sysdeps/nyaux/entry.cpp b/sysdeps/nyaux/entry.cpp
new file mode 100644
index 0000000..4e5b950
--- /dev/null
+++ b/sysdeps/nyaux/entry.cpp
@@ -0,0 +1,14 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <mlibc/elf/startup.h>
+
+extern "C" void __dlapi_enter(uintptr_t *);
+
+extern char **environ;
+
+extern "C" void __mlibc_entry(uintptr_t *entry_stack, int (*main_fn)(int argc, char *argv[], char *env[])) {
+	__dlapi_enter(entry_stack);
+
+	auto result = main_fn(mlibc::entry_stack.argc, mlibc::entry_stack.argv, environ);
+	exit(result);
+}
diff --git a/sysdeps/nyaux/include/abi-bits/access.h b/sysdeps/nyaux/include/abi-bits/access.h
new file mode 120000
index 0000000..cb83931
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/access.h
@@ -0,0 +1 @@
+../../../../abis/linux/access.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/aio.h b/sysdeps/nyaux/include/abi-bits/aio.h
new file mode 120000
index 0000000..79b4168
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/aio.h
@@ -0,0 +1 @@
+../../../../abis/linux/aio.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/auxv.h b/sysdeps/nyaux/include/abi-bits/auxv.h
new file mode 120000
index 0000000..c43f878
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/auxv.h
@@ -0,0 +1 @@
+../../../../abis/linux/auxv.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/blkcnt_t.h b/sysdeps/nyaux/include/abi-bits/blkcnt_t.h
new file mode 120000
index 0000000..0b0ec27
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/blkcnt_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/blkcnt_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/blksize_t.h b/sysdeps/nyaux/include/abi-bits/blksize_t.h
new file mode 120000
index 0000000..7dc8d7c
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/blksize_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/blksize_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/clockid_t.h b/sysdeps/nyaux/include/abi-bits/clockid_t.h
new file mode 120000
index 0000000..6a42da5
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/clockid_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/clockid_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/dev_t.h b/sysdeps/nyaux/include/abi-bits/dev_t.h
new file mode 120000
index 0000000..bca881e
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/dev_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/dev_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/errno.h b/sysdeps/nyaux/include/abi-bits/errno.h
new file mode 120000
index 0000000..6e507de
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/errno.h
@@ -0,0 +1 @@
+../../../../abis/linux/errno.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/fcntl.h b/sysdeps/nyaux/include/abi-bits/fcntl.h
new file mode 120000
index 0000000..463e2c9
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/fcntl.h
@@ -0,0 +1 @@
+../../../../abis/linux/fcntl.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/fd_set.h b/sysdeps/nyaux/include/abi-bits/fd_set.h
new file mode 120000
index 0000000..b43142c
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/fd_set.h
@@ -0,0 +1 @@
+../../../../abis/linux/fd_set.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/fsblkcnt_t.h b/sysdeps/nyaux/include/abi-bits/fsblkcnt_t.h
new file mode 120000
index 0000000..898dfb2
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/fsblkcnt_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/fsblkcnt_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/fsfilcnt_t.h b/sysdeps/nyaux/include/abi-bits/fsfilcnt_t.h
new file mode 120000
index 0000000..791755c
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/fsfilcnt_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/fsfilcnt_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/gid_t.h b/sysdeps/nyaux/include/abi-bits/gid_t.h
new file mode 120000
index 0000000..abce6d6
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/gid_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/gid_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/in.h b/sysdeps/nyaux/include/abi-bits/in.h
new file mode 120000
index 0000000..418d1d5
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/in.h
@@ -0,0 +1 @@
+../../../../abis/linux/in.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/ino_t.h b/sysdeps/nyaux/include/abi-bits/ino_t.h
new file mode 120000
index 0000000..4c20aca
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/ino_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/ino_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/ipc.h b/sysdeps/nyaux/include/abi-bits/ipc.h
new file mode 120000
index 0000000..2c7ffc4
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/ipc.h
@@ -0,0 +1 @@
+../../../../abis/linux/ipc.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/limits.h b/sysdeps/nyaux/include/abi-bits/limits.h
new file mode 120000
index 0000000..6c88db2
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/limits.h
@@ -0,0 +1 @@
+../../../../abis/linux/limits.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/mode_t.h b/sysdeps/nyaux/include/abi-bits/mode_t.h
new file mode 120000
index 0000000..5d78fdf
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/mode_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/mode_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/mqueue.h b/sysdeps/nyaux/include/abi-bits/mqueue.h
new file mode 120000
index 0000000..fa87b07
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/mqueue.h
@@ -0,0 +1 @@
+../../../../abis/linux/mqueue.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/msg.h b/sysdeps/nyaux/include/abi-bits/msg.h
new file mode 120000
index 0000000..f402b49
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/msg.h
@@ -0,0 +1 @@
+../../../../abis/linux/msg.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/nlink_t.h b/sysdeps/nyaux/include/abi-bits/nlink_t.h
new file mode 120000
index 0000000..bb3b625
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/nlink_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/nlink_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/pid_t.h b/sysdeps/nyaux/include/abi-bits/pid_t.h
new file mode 120000
index 0000000..baa90f6
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/pid_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/pid_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/poll.h b/sysdeps/nyaux/include/abi-bits/poll.h
new file mode 120000
index 0000000..8ea6a0a
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/poll.h
@@ -0,0 +1 @@
+../../../../abis/linux/poll.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/resource.h b/sysdeps/nyaux/include/abi-bits/resource.h
new file mode 120000
index 0000000..88d7402
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/resource.h
@@ -0,0 +1 @@
+../../../../abis/linux/resource.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/riscv-hwprobe.h b/sysdeps/nyaux/include/abi-bits/riscv-hwprobe.h
new file mode 120000
index 0000000..e651325
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/riscv-hwprobe.h
@@ -0,0 +1 @@
+../../../../abis/linux/riscv-hwprobe.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/rlim_t.h b/sysdeps/nyaux/include/abi-bits/rlim_t.h
new file mode 120000
index 0000000..e92eb5f
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/rlim_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/rlim_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sa_family_t.h b/sysdeps/nyaux/include/abi-bits/sa_family_t.h
new file mode 120000
index 0000000..a75e229
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sa_family_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/sa_family_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sched_param.h b/sysdeps/nyaux/include/abi-bits/sched_param.h
new file mode 120000
index 0000000..fa4eb2c
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sched_param.h
@@ -0,0 +1 @@
+../../../../abis/linux/sched_param.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/seek-whence.h b/sysdeps/nyaux/include/abi-bits/seek-whence.h
new file mode 120000
index 0000000..df7bccf
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/seek-whence.h
@@ -0,0 +1 @@
+../../../../abis/linux/seek-whence.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sem.h b/sysdeps/nyaux/include/abi-bits/sem.h
new file mode 120000
index 0000000..699d7af
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sem.h
@@ -0,0 +1 @@
+../../../../abis/linux/sem.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/shm.h b/sysdeps/nyaux/include/abi-bits/shm.h
new file mode 120000
index 0000000..067d8c4
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/shm.h
@@ -0,0 +1 @@
+../../../../abis/linux/shm.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sig-limits.h b/sysdeps/nyaux/include/abi-bits/sig-limits.h
new file mode 120000
index 0000000..f829738
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sig-limits.h
@@ -0,0 +1 @@
+../../../../abis/linux/sig-limits.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sigevent.h b/sysdeps/nyaux/include/abi-bits/sigevent.h
new file mode 120000
index 0000000..83d069b
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sigevent.h
@@ -0,0 +1 @@
+../../../../abis/linux/sigevent.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/signal.h b/sysdeps/nyaux/include/abi-bits/signal.h
new file mode 120000
index 0000000..4dcb0b7
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/signal.h
@@ -0,0 +1 @@
+../../../../abis/linux/signal.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sigset_t.h b/sysdeps/nyaux/include/abi-bits/sigset_t.h
new file mode 120000
index 0000000..8b8e81e
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sigset_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/sigset_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sigval.h b/sysdeps/nyaux/include/abi-bits/sigval.h
new file mode 120000
index 0000000..ccd43a5
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sigval.h
@@ -0,0 +1 @@
+../../../../abis/linux/sigval.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/sockaddr_storage.h b/sysdeps/nyaux/include/abi-bits/sockaddr_storage.h
new file mode 120000
index 0000000..b374865
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/sockaddr_storage.h
@@ -0,0 +1 @@
+../../../../abis/linux/sockaddr_storage.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/socket.h b/sysdeps/nyaux/include/abi-bits/socket.h
new file mode 120000
index 0000000..f1dc016
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/socket.h
@@ -0,0 +1 @@
+../../../../abis/linux/socket.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/socklen_t.h b/sysdeps/nyaux/include/abi-bits/socklen_t.h
new file mode 120000
index 0000000..41f3b11
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/socklen_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/socklen_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/stat.h b/sysdeps/nyaux/include/abi-bits/stat.h
new file mode 120000
index 0000000..1f63b41
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/stat.h
@@ -0,0 +1 @@
+../../../../abis/linux/stat.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/statvfs.h b/sysdeps/nyaux/include/abi-bits/statvfs.h
new file mode 120000
index 0000000..1fc80c2
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/statvfs.h
@@ -0,0 +1 @@
+../../../../abis/linux/statvfs.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/suseconds_t.h b/sysdeps/nyaux/include/abi-bits/suseconds_t.h
new file mode 120000
index 0000000..9ed6597
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/suseconds_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/suseconds_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/termios.h b/sysdeps/nyaux/include/abi-bits/termios.h
new file mode 120000
index 0000000..ee8f0b0
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/termios.h
@@ -0,0 +1 @@
+../../../../abis/linux/termios.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/time.h b/sysdeps/nyaux/include/abi-bits/time.h
new file mode 120000
index 0000000..2a02625
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/time.h
@@ -0,0 +1 @@
+../../../../abis/linux/time.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/uid_t.h b/sysdeps/nyaux/include/abi-bits/uid_t.h
new file mode 120000
index 0000000..b306777
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/uid_t.h
@@ -0,0 +1 @@
+../../../../abis/linux/uid_t.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/utmp-defines.h b/sysdeps/nyaux/include/abi-bits/utmp-defines.h
new file mode 120000
index 0000000..8617643
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/utmp-defines.h
@@ -0,0 +1 @@
+../../../../abis/linux/utmp-defines.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/utmpx.h b/sysdeps/nyaux/include/abi-bits/utmpx.h
new file mode 120000
index 0000000..c6a2677
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/utmpx.h
@@ -0,0 +1 @@
+../../../../abis/linux/utmpx.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/utsname.h b/sysdeps/nyaux/include/abi-bits/utsname.h
new file mode 120000
index 0000000..b285754
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/utsname.h
@@ -0,0 +1 @@
+../../../../abis/linux/utsname.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/vm-flags.h b/sysdeps/nyaux/include/abi-bits/vm-flags.h
new file mode 120000
index 0000000..bbe258c
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/vm-flags.h
@@ -0,0 +1 @@
+../../../../abis/linux/vm-flags.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/abi-bits/wait.h b/sysdeps/nyaux/include/abi-bits/wait.h
new file mode 120000
index 0000000..feb2840
--- /dev/null
+++ b/sysdeps/nyaux/include/abi-bits/wait.h
@@ -0,0 +1 @@
+../../../../abis/linux/wait.h
\ No newline at end of file
diff --git a/sysdeps/nyaux/include/bits/syscall.h b/sysdeps/nyaux/include/bits/syscall.h
new file mode 100644
index 0000000..680bb8c
--- /dev/null
+++ b/sysdeps/nyaux/include/bits/syscall.h
@@ -0,0 +1,100 @@
+#ifndef _MLIBC_SYSCALL_H
+#define _MLIBC_SYSCALL_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef long __sc_word_t;
+
+/* These functions are implemented in arch-syscall.cpp. */
+__sc_word_t __do_syscall0(long);
+__sc_word_t __do_syscall1(long, __sc_word_t);
+__sc_word_t __do_syscall2(long, __sc_word_t, __sc_word_t);
+__sc_word_t __do_syscall3(long, __sc_word_t, __sc_word_t, __sc_word_t);
+__sc_word_t __do_syscall4(long, __sc_word_t, __sc_word_t, __sc_word_t, __sc_word_t);
+__sc_word_t __do_syscall5(long, __sc_word_t, __sc_word_t, __sc_word_t, __sc_word_t,
+		__sc_word_t);
+__sc_word_t __do_syscall6(long, __sc_word_t, __sc_word_t, __sc_word_t, __sc_word_t,
+		__sc_word_t, __sc_word_t);
+__sc_word_t __do_syscall7(long, __sc_word_t, __sc_word_t, __sc_word_t, __sc_word_t,
+		__sc_word_t, __sc_word_t, __sc_word_t);
+long __do_syscall_ret(unsigned long);
+
+#ifdef __cplusplus
+extern "C++" {
+
+/* Defining a syscall as a macro is more problematic in C++, since there's a high chance of
+ * a name collision e.g foo.syscall() or foo::syscall.
+ */
+inline long syscall(long n) {
+	return __do_syscall_ret(__do_syscall0(n));
+}
+template<typename Arg0>
+long syscall(long n, Arg0 a0) {
+	return __do_syscall_ret(__do_syscall1(n, (long)a0));
+}
+template<typename Arg0, typename Arg1>
+long syscall(long n, Arg0 a0, Arg1 a1) {
+	return __do_syscall_ret(__do_syscall2(n, (long)a0, (long)a1));
+}
+template<typename Arg0, typename Arg1, typename Arg2>
+long syscall(long n, Arg0 a0, Arg1 a1, Arg2 a2) {
+	return __do_syscall_ret(__do_syscall3(n, (long)a0, (long)a1, (long)a2));
+}
+template<typename Arg0, typename Arg1, typename Arg2, typename Arg3>
+long syscall(long n, Arg0 a0, Arg1 a1, Arg2 a2, Arg3 a3) {
+	return __do_syscall_ret(__do_syscall4(n, (long)a0, (long)a1, (long)a2, (long)a3));
+}
+template<typename Arg0, typename Arg1, typename Arg2, typename Arg3, typename Arg4>
+long syscall(long n, Arg0 a0, Arg1 a1, Arg2 a2, Arg3 a3, Arg4 a4) {
+	return __do_syscall_ret(__do_syscall5(n, (long)a0, (long)a1, (long)a2, (long)a3, (long)a4));
+}
+template<typename Arg0, typename Arg1, typename Arg2, typename Arg3, typename Arg4, typename Arg5>
+long syscall(long n, Arg0 a0, Arg1 a1, Arg2 a2, Arg3 a3, Arg4 a4, Arg5 a5) {
+	return __do_syscall_ret(__do_syscall6(n, (long)a0, (long)a1, (long)a2, (long)a3, (long)a4, (long)a5));
+}
+template<typename Arg0, typename Arg1, typename Arg2, typename Arg3, typename Arg4, typename Arg5, typename Arg6>
+long syscall(long n, Arg0 a0, Arg1 a1, Arg2 a2, Arg3 a3, Arg4 a4, Arg5 a5, Arg6 a6) {
+	return __do_syscall_ret(__do_syscall7(n, (long)a0, (long)a1, (long)a2, (long)a3, (long)a4, (long)a5, (long)a6));
+}
+
+} /* extern C++ */
+#else
+
+/*
+ * Variadic macros are not supported in C89.
+ * glibc implements syscall() as a variadic function, which we've ruled out.
+ * musl uses them without checking the C standard in use. So suppressing
+ * the check here seems reasonable.
+ */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wvariadic-macros"
+
+/* These syscall macros were copied from musl. */
+#define __scc(x) ((__sc_word_t)(x))
+#define __syscall0(n) __do_syscall0(n)
+#define __syscall1(n,a) __do_syscall1(n,__scc(a))
+#define __syscall2(n,a,b) __do_syscall2(n,__scc(a),__scc(b))
+#define __syscall3(n,a,b,c) __do_syscall3(n,__scc(a),__scc(b),__scc(c))
+#define __syscall4(n,a,b,c,d) __do_syscall4(n,__scc(a),__scc(b),__scc(c),__scc(d))
+#define __syscall5(n,a,b,c,d,e) __do_syscall5(n,__scc(a),__scc(b),__scc(c),__scc(d),__scc(e))
+#define __syscall6(n,a,b,c,d,e,f) __do_syscall6(n,__scc(a),__scc(b),__scc(c),__scc(d),__scc(e),__scc(f))
+#define __syscall7(n,a,b,c,d,e,f,g) __do_syscall7(n,__scc(a),__scc(b),__scc(c),__scc(d),__scc(e),__scc(f),__scc(g))
+#define __SYSCALL_NARGS_X(a,b,c,d,e,f,g,h,n,...) n
+#define __SYSCALL_NARGS(...) __SYSCALL_NARGS_X(__VA_ARGS__,7,6,5,4,3,2,1,0,)
+#define __SYSCALL_CONCAT_X(a,b) a##b
+#define __SYSCALL_CONCAT(a,b) __SYSCALL_CONCAT_X(a,b)
+#define __SYSCALL_DISP(b,...) __SYSCALL_CONCAT(b,__SYSCALL_NARGS(__VA_ARGS__))(__VA_ARGS__)
+#define __syscall(...) __SYSCALL_DISP(__syscall,__VA_ARGS__)
+#define syscall(...) __do_syscall_ret(__syscall(__VA_ARGS__))
+
+#pragma GCC diagnostic pop
+
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* _MLIBC_SYSCALL_H */
diff --git a/sysdeps/nyaux/include/mlibc/sysdeps.hpp b/sysdeps/nyaux/include/mlibc/sysdeps.hpp
new file mode 100644
index 0000000..3f75802
--- /dev/null
+++ b/sysdeps/nyaux/include/mlibc/sysdeps.hpp
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <mlibc/sysdep-signatures.hpp>
+
+namespace mlibc {
+
+struct DemoSysdepTags :
+	LibcPanic,
+	LibcLog,
+	Isatty,
+	Write,
+	TcbSet,
+	AnonAllocate,
+	AnonFree,
+	Seek,
+	Exit,
+	Close,
+	FutexWake,
+	FutexWait,
+	Read,
+	Open,
+	VmMap,
+	VmUnmap,
+	ClockGet
+{};
+
+template<typename Tag>
+using Sysdeps = SysdepOf<DemoSysdepTags, Tag>;
+
+struct SysdepTraits {
+	static constexpr bool usesRtNetlink = false;
+};
+
+} // namespace mlibc
diff --git a/sysdeps/nyaux/meson.build b/sysdeps/nyaux/meson.build
new file mode 100644
index 0000000..17c9b4a
--- /dev/null
+++ b/sysdeps/nyaux/meson.build
@@ -0,0 +1,121 @@
+# ANCHOR: supported-options
+sysdep_supported_options = {
+	'internal': true,
+	'ansi': true,
+	'posix': true,
+}
+# ANCHOR_END: supported-options
+
+# ANCHOR: sources-and-includes
+rtld_sources += files(
+	'sysdeps.cpp',
+	'syscall.cpp',
+)
+rtld_include_dirs += include_directories('include')
+
+libc_sources += files(
+	'entry.cpp',
+	'sysdeps.cpp',
+	'syscall.cpp',
+)
+libc_include_dirs += include_directories('include')
+# ANCHOR_END: sources-and-includes
+
+# ANCHOR: abi-includes
+if not no_headers
+	install_headers(
+		'include/abi-bits/auxv.h',
+		'include/abi-bits/blkcnt_t.h',
+		'include/abi-bits/blksize_t.h',
+		'include/abi-bits/clockid_t.h',
+		'include/abi-bits/dev_t.h',
+		'include/abi-bits/errno.h',
+		'include/abi-bits/fcntl.h',
+		'include/abi-bits/gid_t.h',
+		'include/abi-bits/ino_t.h',
+		'include/abi-bits/limits.h',
+		'include/abi-bits/mode_t.h',
+		'include/abi-bits/nlink_t.h',
+		'include/abi-bits/pid_t.h',
+		'include/abi-bits/seek-whence.h',
+		'include/abi-bits/signal.h',
+		'include/abi-bits/stat.h',
+		'include/abi-bits/uid_t.h',
+		'include/abi-bits/vm-flags.h',
+		'include/abi-bits/wait.h',
+		'include/abi-bits/riscv-hwprobe.h',
+		'include/abi-bits/sigevent.h',
+		'include/abi-bits/sigval.h',
+		'include/abi-bits/sa_family_t.h',
+		'include/abi-bits/sockaddr_storage.h',
+		'include/abi-bits/sig-limits.h',
+		'include/abi-bits/suseconds_t.h',
+		'include/abi-bits/access.h',
+		'include/abi-bits/socklen_t.h',
+		'include/abi-bits/socket.h',
+		'include/abi-bits/poll.h',
+		'include/abi-bits/resource.h',
+		'include/abi-bits/in.h',
+		'include/abi-bits/rlim_t.h',
+		'include/abi-bits/utsname.h',
+		'include/abi-bits/fd_set.h',
+		'include/abi-bits/sem.h',
+		'include/abi-bits/time.h',
+		'include/abi-bits/ipc.h',
+		'include/abi-bits/statvfs.h',
+		'include/abi-bits/fsblkcnt_t.h',
+		'include/abi-bits/fsfilcnt_t.h',
+		'include/abi-bits/shm.h',
+		'include/abi-bits/termios.h',
+		'include/abi-bits/msg.h',
+		'include/abi-bits/mqueue.h',
+		'include/abi-bits/utmp-defines.h',
+		'include/abi-bits/utmpx.h',
+		'include/abi-bits/sigset_t.h',
+		'include/abi-bits/aio.h',
+		'include/abi-bits/sched_param.h',
+		subdir: 'abi-bits',
+		follow_symlinks: true
+	)
+endif
+# ANCHOR_END: abi-includes
+
+# ANCHOR: crt1
+if not headers_only
+	crt = custom_target('crt1',
+		build_by_default: true,
+		command: c_compiler.cmd_array() + ['-c', '-o', '@OUTPUT@', '@INPUT@'],
+		input: 'crt1.S',
+		output: 'crt1.o',
+		install: true,
+		install_dir: get_option('libdir')
+	)
+
+    crt_pie = custom_target('Scrt1',
+		build_by_default: true,
+		command: c_compiler.cmd_array() + ['-c', '-o', '@OUTPUT@', '@INPUT@'],
+		input: 'crt1.S',
+		output: 'Scrt1.o',
+		install: true,
+		install_dir: get_option('libdir')
+	)
+
+	custom_target('crti',
+		build_by_default: true,
+		command: c_compiler.cmd_array() + ['-c', '-o', '@OUTPUT@', '@INPUT@'],
+		input: 'crti.S',
+		output: 'crti.o',
+		install: true,
+		install_dir: get_option('libdir')
+	)
+
+	custom_target('crtn',
+		build_by_default: true,
+		command: c_compiler.cmd_array() + ['-c', '-o', '@OUTPUT@', '@INPUT@'],
+		input: 'crtn.S',
+		output: 'crtn.o',
+		install: true,
+		install_dir: get_option('libdir')
+	)
+endif
+# ANCHOR_END: crt1
diff --git a/sysdeps/nyaux/syscall.cpp b/sysdeps/nyaux/syscall.cpp
new file mode 100644
index 0000000..a7dae56
--- /dev/null
+++ b/sysdeps/nyaux/syscall.cpp
@@ -0,0 +1,82 @@
+#include <bits/syscall.h>
+#include <errno.h>
+
+extern "C" long __do_syscall_ret(unsigned long ret) {
+	return ret;
+}
+
+using sc_word_t = __sc_word_t;
+
+sc_word_t __do_syscall0(long sc) {
+	sc_word_t ret;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc)
+			: "rcx", "r11", "memory");
+	return ret;
+}
+
+sc_word_t __do_syscall1(long sc,
+		sc_word_t arg1) {
+	sc_word_t ret;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc), "D"(arg1)
+			: "rcx", "r11", "memory");
+	return ret;
+}
+
+sc_word_t __do_syscall2(long sc,
+		sc_word_t arg1, sc_word_t arg2) {
+	sc_word_t ret;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc), "D"(arg1), "S"(arg2)
+			: "rcx", "r11", "memory");
+	return ret;
+}
+
+sc_word_t __do_syscall3(long sc,
+		sc_word_t arg1, sc_word_t arg2, sc_word_t arg3) {
+	sc_word_t ret;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc), "D"(arg1), "S"(arg2), "d"(arg3)
+			: "rcx", "r11", "memory");
+	return ret;
+}
+
+sc_word_t __do_syscall4(long sc,
+		sc_word_t arg1, sc_word_t arg2, sc_word_t arg3,
+		sc_word_t arg4) {
+	sc_word_t ret;
+	register sc_word_t arg4_reg asm("r10") = arg4;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc), "D"(arg1), "S"(arg2), "d"(arg3),
+				"r"(arg4_reg)
+			: "rcx", "r11", "memory");
+	return ret;
+}
+
+sc_word_t __do_syscall5(long sc,
+		sc_word_t arg1, sc_word_t arg2, sc_word_t arg3,
+		sc_word_t arg4, sc_word_t arg5) {
+	sc_word_t ret;
+	register sc_word_t arg4_reg asm("r10") = arg4;
+	register sc_word_t arg5_reg asm("r8") = arg5;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc), "D"(arg1), "S"(arg2), "d"(arg3),
+				"r"(arg4_reg), "r"(arg5_reg)
+			: "rcx", "r11", "memory");
+	return ret;
+}
+
+sc_word_t __do_syscall6(long sc,
+		sc_word_t arg1, sc_word_t arg2, sc_word_t arg3,
+		sc_word_t arg4, sc_word_t arg5, sc_word_t arg6) {
+	sc_word_t ret;
+	register sc_word_t arg4_reg asm("r10") = arg4;
+	register sc_word_t arg5_reg asm("r8") = arg5;
+	register sc_word_t arg6_reg asm("r9") = arg6;
+	asm volatile ("syscall" : "=a"(ret)
+			: "a"(sc), "D"(arg1), "S"(arg2), "d"(arg3),
+				"r"(arg4_reg), "r"(arg5_reg), "r"(arg6_reg)
+			: "rcx", "r11", "memory");
+	return ret;
+}
diff --git a/sysdeps/nyaux/sysdeps.cpp b/sysdeps/nyaux/sysdeps.cpp
new file mode 100644
index 0000000..cc0fb96
--- /dev/null
+++ b/sysdeps/nyaux/sysdeps.cpp
@@ -0,0 +1,106 @@
+#include "mlibc/tcb.hpp"
+#include <abi-bits/errno.h>
+#include <bits/ensure.h>
+#include <bits/syscall.h>
+#include <mlibc/all-sysdeps.hpp>
+#include <string.h>
+
+#define SYS_EXIT 0
+#define SYS_WRITE 1
+#define SYS_MMAP 2
+
+// ANCHOR: stub
+#define STUB()                                                                                     \
+	({                                                                                             \
+		__ensure(!"STUB function was called");                                                     \
+		__builtin_unreachable();                                                                   \
+	})
+// ANCHOR_END: stub
+
+namespace mlibc {
+
+void Sysdeps<LibcPanic>::operator()() {
+	sysdep<LibcLog>("!!! mlibc panic !!!");
+	sysdep<Exit>(-1);
+	__builtin_trap();
+}
+
+void Sysdeps<LibcLog>::operator()(const char *msg) {
+	ssize_t unused;
+	sysdep<Write>(2, msg, strlen(msg), &unused);
+}
+
+int Sysdeps<Isatty>::operator()(int fd) {
+	(void)fd;
+	// this returns ENOTTY when it is not a tty, but we do not have a proper implementation
+	// so always return that a file is a tty
+	return 0;
+}
+
+int Sysdeps<Write>::operator()(int fd, void const *buf, size_t size, ssize_t *ret) {
+	*ret = syscall(SYS_WRITE, fd, buf, size);
+	// this can never fail in the demo os
+	return 0;
+}
+
+int Sysdeps<TcbSet>::operator()(void *pointer) {
+	uintptr_t thread_data = reinterpret_cast<uintptr_t>(pointer) + sizeof(Tcb);
+	// this can never fail in the demo os
+	return 0;
+}
+
+int Sysdeps<AnonAllocate>::operator()(size_t size, void **pointer) {
+	auto out = syscall(
+	    SYS_MMAP, nullptr, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0
+	);
+	*pointer = (void *)out;
+	if (*pointer == MAP_FAILED)
+		return ENOMEM; // the syscall does not return a proper errno, so use ENOMEM
+	return 0;
+}
+
+int Sysdeps<AnonFree>::operator()(void *, unsigned long) {
+	return 0;
+} // no-op
+
+int Sysdeps<Seek>::operator()(int, off_t, int, off_t *) {
+	return ESPIPE; // no proper file implementation, everything is a tty so return ESPIPE
+}
+
+void Sysdeps<Exit>::operator()(int status) {
+	syscall(SYS_EXIT, status);
+	__builtin_unreachable();
+}
+
+// ANCHOR: stubbed-sysdep
+int Sysdeps<Close>::operator()(int) {
+	STUB();
+}
+// ANCHOR_END: stubbed-sysdep
+
+int Sysdeps<FutexWake>::operator()(int *, bool) {
+	STUB();
+}
+int Sysdeps<FutexWait>::operator()(int *, int, timespec const *) {
+	STUB();
+}
+int Sysdeps<Read>::operator()(int, void *, unsigned long, long *) {
+	STUB();
+}
+int Sysdeps<Open>::operator()(const char *, int, unsigned int, int *) {
+	STUB();
+}
+int Sysdeps<VmMap>::operator()(void *, size_t, int, int, int, off_t, void **) {
+	STUB();
+}
+int Sysdeps<VmUnmap>::operator()(void *, size_t) {
+	STUB();
+}
+int Sysdeps<ClockGet>::operator()(int, time_t *, long *) {
+	STUB();
+}
+
+} // namespace mlibc
+
+ // i fixed the patch liz :3 since we in container now
