diff --git ncurses-clean/c++/Makefile.in ncurses-workdir/c++/Makefile.in
index 99f21da..5db41d8 100644
--- ncurses-clean/c++/Makefile.in
+++ ncurses-workdir/c++/Makefile.in
@@ -118,7 +118,7 @@ SHLIB_LIST	= $(SHLIB_DIRS) \
 		-l@FORM_NAME@@USE_LIB_SUFFIX@ \
 		-l@MENU_NAME@@USE_LIB_SUFFIX@ \
 		-l@PANEL_NAME@@USE_LIB_SUFFIX@ \
-		-lncurses@USE_LIB_SUFFIX@ @SHLIB_LIST@
+		-lncurses@USE_LIB_SUFFIX@
 
 LIBROOT		= ncurses++
 
@@ -157,7 +157,7 @@ LDFLAGS_SHARED	= $(TEST_LDFLAGS) $(CFLAGS_SHARED) @LD_SHARED_OPTS@
 LDFLAGS_DEFAULT	= $(LINK_@DFT_UPR_MODEL@) $(LDFLAGS_@DFT_UPR_MODEL@)
 
 # flags for library built by this makefile
-LDFLAGS		= $(TEST_ARGS) @LDFLAGS@ \
+LDFLAGS		= @LDFLAGS@ \
 	@LD_MODEL@ $(TEST_LIBS) @LIBS@ $(CXXLIBS)
 
 AUTO_SRC	= \
@@ -182,7 +182,7 @@ AUTO_SRC	= \
 all \
 libs ::		$(AUTO_SRC) $(LIBRARIES)
 
-@MAKE_TESTS@all ::		demo$x
+#@MAKE_TESTS@all ::		demo$x
 
 depend :
 
@@ -256,20 +256,20 @@ $(INCDIR)/panel.h :
 
 ###############################################################################
 
-OBJS_DEMO = $(MODEL)/demo$o
+#OBJS_DEMO = $(MODEL)/demo$o
 
-$(MODEL)/demo$o : $(srcdir)/demo.cc \
-		$(HEADER_DEPS) \
-		$(cursesf_h) \
-		$(cursesm_h) \
-		$(cursesapp_h)
-	@echo 'compiling demo (obj_s)'
-	@( cd $(MODEL) && $(LIBTOOL_COMPILE) $(CXX) $(CFLAGS_DEFAULT) -c $(top_srcdir)/c++/demo.cc -o $@ )
+#$(MODEL)/demo$o : $(srcdir)/demo.cc \
+#		$(HEADER_DEPS) \
+#		$(cursesf_h) \
+#		$(cursesm_h) \
+#		$(cursesapp_h)
+#	@echo 'compiling demo (obj_s)'
+#	@( cd $(MODEL) && $(LIBTOOL_COMPILE) $(CXX) $(CFLAGS_DEFAULT) -c $(top_srcdir)/c++/demo.cc -o $@ )
 
-demo$x:	$(OBJS_DEMO) \
-	$(LIBRARIES)  \
-	@TEST_DEPS@
-	@ECHO_LD@ $(LINK) -o $@ $(OBJS_DEMO) $(LDFLAGS_DEFAULT)
+#demo$x:	$(OBJS_DEMO) \
+#	$(LIBRARIES)  \
+#	@TEST_DEPS@
+#	@ECHO_LD@ $(LINK) -o $@ $(OBJS_DEMO) $(LDFLAGS_DEFAULT)
 
 etip.h:	$(srcdir)/etip.h.in $(srcdir)/edit_cfg.sh
 	cp $(srcdir)/etip.h.in $@
diff --git ncurses-clean/configure ncurses-workdir/configure
index 488d93f..55ed245 100755
--- ncurses-clean/configure
+++ ncurses-workdir/configure
@@ -3520,7 +3520,7 @@ echo $ECHO_N "checking if $CXX works... $ECHO_C" >&6
 
 		save_CPPFLAGS="$CPPFLAGS"
 		eval cf_includedir=${includedir}
-		CPPFLAGS="$CPPFLAGS -I${cf_includedir}"
+		CPPFLAGS="$CPPFLAGS"
 
 cat >"conftest.$ac_ext" <<_ACEOF
 #line 3526 "configure"
@@ -6925,6 +6925,10 @@ CF_EOF
 		LINK_PROGS="$SHELL ${rel_builddir}/mk_prog.sh"
 		LINK_TESTS="$SHELL ${rel_builddir}/mk_prog.sh"
 		;;
+	(nyaux*)
+		CC_SHARED_OPTS='-fPIC'
+		MK_SHARED_LIB='${CC} -shared -o $@'
+		;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
diff --git ncurses-clean/form/Makefile.in ncurses-workdir/form/Makefile.in
index a950a3e..9b092c0 100644
--- ncurses-clean/form/Makefile.in
+++ ncurses-workdir/form/Makefile.in
@@ -110,7 +110,7 @@ LINK		= $(LIBTOOL_LINK)
 LDFLAGS		= @LDFLAGS@ @LD_MODEL@ @LIBS@
 
 SHLIB_DIRS	= -L../lib
-SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@ @SHLIB_LIST@
+SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@
 
 RPATH_LIST	= @RPATH_LIST@
 RESULTING_SYMS	= @RESULTING_SYMS@
diff --git ncurses-clean/misc/gen-pkgconfig.in ncurses-workdir/misc/gen-pkgconfig.in
index 89a5cd4..07d94d1 100644
--- ncurses-clean/misc/gen-pkgconfig.in
+++ ncurses-workdir/misc/gen-pkgconfig.in
@@ -83,7 +83,7 @@ if [ "$includedir" != "/usr/include" ]; then
 fi
 
 lib_flags=
-for opt in -L$libdir @EXTRA_PKG_LDFLAGS@ @LIBS@
+for opt in -L$libdir @LIBS@
 do
 	case $opt in
 	-l*) # LIBS is handled specially below
diff --git ncurses-clean/misc/ncurses-config.in ncurses-workdir/misc/ncurses-config.in
index c9f1cf8..51798da 100644
--- ncurses-clean/misc/ncurses-config.in
+++ ncurses-workdir/misc/ncurses-config.in
@@ -101,7 +101,7 @@ fi
 # There is no portable way to find the list of standard library directories.
 # Require a POSIX shell anyway, to keep this simple.
 lib_flags=
-for opt in -L$libdir @EXTRA_PKG_LDFLAGS@ $LIBS
+for opt in -L$libdir $LIBS
 do
 	case $opt in
 	-specs*) # ignore linker specs-files which were used to build library
diff --git ncurses-clean/ncurses/base/lib_getch.c ncurses-workdir/ncurses/base/lib_getch.c
index 8e47679..bbf7748 100644
--- ncurses-clean/ncurses/base/lib_getch.c
+++ ncurses-workdir/ncurses/base/lib_getch.c
@@ -453,17 +453,19 @@ _nc_wgetch(WINDOW *win,
 #endif
 
     T((T_CALLED("_nc_wgetch(%p)"), (void *) win));
-
+	
     *result = 0;
 
     sp = _nc_screen_of(win);
     if (win == 0 || sp == 0) {
 	returnCode(ERR);
     }
-
+	
     if (cooked_key_in_fifo()) {
+		
 	recur_wrefresh(win);
 	*result = fifo_pull(sp);
+	
 	returnCode(*result >= KEY_MIN ? KEY_CODE_YES : OK);
     }
 #ifdef NCURSES_WGETCH_EVENTS
@@ -483,7 +485,7 @@ _nc_wgetch(WINDOW *win,
 	!IsCbreak(sp) &&
 	!sp->_called_wgetch) {
 	char buf[MAXCOLUMNS], *bufp;
-
+	
 	TR(TRACE_IEVENT, ("filling queue in cooked mode"));
 
 	/* ungetch in reverse order */
@@ -505,6 +507,7 @@ _nc_wgetch(WINDOW *win,
 	} else
 #endif
 	    *result = fifo_pull(sp);
+		
 	returnCode(*result >= KEY_MIN ? KEY_CODE_YES : OK);
     }
 
@@ -512,11 +515,12 @@ _nc_wgetch(WINDOW *win,
 	_nc_keypad(sp, win->_use_keypad);
 
     recur_wrefresh(win);
-
+	
     if (win->_notimeout || (win->_delay >= 0) || (IsCbreak(sp) > 1)) {
+		
 	if (head == -1) {	/* fifo is empty */
 	    int delay;
-
+		
 	    TR(TRACE_IEVENT, ("timed delay in wgetch()"));
 	    if (IsCbreak(sp) > 1)
 		delay = (IsCbreak(sp) - 1) * 100;
@@ -543,8 +547,9 @@ _nc_wgetch(WINDOW *win,
 	    }
 	}
 	/* else go on to read data available */
+	
     }
-
+	
     if (win->_use_keypad) {
 	/*
 	 * This is tricky.  We only want to get special-key
@@ -598,8 +603,8 @@ _nc_wgetch(WINDOW *win,
 	    fifo_push(sp EVENTLIST_2nd(evl));
 	ch = fifo_pull(sp);
     }
-
     if (ch == ERR) {
+		
       check_sigwinch:
 #if USE_SIZECHANGE
 	if (_nc_handle_sigwinch(sp)) {
@@ -720,17 +725,20 @@ kgetch(SCREEN *sp, bool forever EVENTLIST_2nd(_nc_eventlist * evl))
     TRIES *ptr;
     int ch = 0;
     int timeleft = forever ? 9999999 : GetEscdelay(sp);
-
+	
     TR(TRACE_IEVENT, ("kgetch() called"));
 
     ptr = sp->_keytry;
 
     for (;;) {
 	if (cooked_key_in_fifo() && sp->_fifo[head] >= KEY_MIN) {
+
 	    break;
 	} else if (!raw_key_in_fifo()) {
+		
 	    ch = fifo_push(sp EVENTLIST_2nd(evl));
 	    if (ch == ERR) {
+			fprintf(stderr, "ncurses: kill me\r\n");
 		peek = head;	/* the keys stay uninterpreted */
 		return ERR;
 	    }
diff --git ncurses-clean/panel/Makefile.in ncurses-workdir/panel/Makefile.in
index 380f80c..a77ef15 100644
--- ncurses-clean/panel/Makefile.in
+++ ncurses-workdir/panel/Makefile.in
@@ -112,7 +112,7 @@ LINK		= $(LIBTOOL_LINK)
 LDFLAGS		= @LDFLAGS@ @LD_MODEL@ @LIBS@
 
 SHLIB_DIRS	= -L../lib
-SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@ @SHLIB_LIST@
+SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@
 
 RPATH_LIST	= @RPATH_LIST@
 RESULTING_SYMS	= @RESULTING_SYMS@
diff --git ncurses-clean/progs/Makefile.in ncurses-workdir/progs/Makefile.in
index 218e64e..160d308 100644
--- ncurses-clean/progs/Makefile.in
+++ ncurses-workdir/progs/Makefile.in
@@ -79,7 +79,7 @@ LIBTOOL_INSTALL	= @LIB_INSTALL@
 LIBTOOL_UNINSTALL = @LIB_UNINSTALL@
 
 INSTALL		= @INSTALL@
-INSTALL_PROG	= @INSTALL_PROGRAM@ @INSTALL_OPT_S@
+INSTALL_PROG	= @INSTALL_PROGRAM@
 transform	= @program_transform_name@
 
 AWK		= @AWK@
